stdlib/type.mli

```diff
@@ -35,3 +35,100 @@ type (_, _) eq = Equal: ('a, 'a) eq
 
     At runtime, this function simply returns its second argument unchanged.
 *)
+
+(** {1:identifiers Type identifiers} *)
+
+(** Type identifiers.
+
+    A type identifier is a value that denotes a type. Given two type
+    identifiers, they can be tested for {{!Id.typed_equal}equality} to
+    prove they denote the same type. Note that:
+
+    - Unequal identifiers do not imply unequal types: a given type can be
+      denoted by more than one identifier.
+    - Type identifiers can be marshalled, but they get a new, distinct,
+      identity on unmarshalling, so the equalities are lost.
+
+    See an {{!Id.example}example} of use. *)
+module Id : sig
+
+  (** {1:ids Type identifiers} *)
+
+  type 'a t
```

Comment:
```suggestion
  type !'a t
```
(Injectivity annotations for type constructors that are necessarily injective are now [standard style in the standard library](https://github.com/ocaml/ocaml/pull/9781))

Reply:
The type system thinks this is invariant: 

```
29 |   type !'a t = (module ID with type t = 'a)
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: In this definition, expected parameter variances are not satisfied.
       The 1st type parameter was expected to be injective invariant,
       but it is invariant.
```

Reply:
That's interesting.  The injectivity annotation was accepted in OCaml 5.0.0.

Reply:
See #11869.  There's no reason for the issue to block the merge of this PR; we can always add an injectivity annotation later if necessary.